experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1113 @@
+                    MappingKind::Upper | MappingKind::Title => self
+                        .full_upper_or_title_special_case(
+                            c,
+                            context,
+                            locale,
+                            kind == MappingKind::Title,
+                        ),
+                } {
+                    return special;
+                }
+            }
+            if self.exceptions.has_slot(idx, ExceptionSlot::FullMappings) {
+                let mapped_string = self.exceptions.full_mapping_string(idx, kind);
+                if !mapped_string.is_empty() {
+                    return FullMappingResult::String(mapped_string);
+                }
+            }
+            if data.is_relevant_to(kind) && self.exceptions.has_slot(idx, ExceptionSlot::Delta) {
+                let mapped = c as i32 + self.exceptions.delta(idx);
+                let mapped = char::from_u32(mapped as u32).expect("Checked in validate()");
+                return FullMappingResult::CodePoint(mapped);
+            }
+            if kind == MappingKind::Fold && self.exceptions.no_simple_case_folding(idx) {
+                return FullMappingResult::CodePoint(c);
+            }
+
+            if let Some(slot_char) = self.exceptions.slot_char_for_kind(idx, kind) {
+                FullMappingResult::CodePoint(slot_char)
+            } else {
+                FullMappingResult::CodePoint(c)
+            }
+        }
+    }
+
+    pub(crate) fn to_full_lower(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Lower)
+    }
+
+    pub(crate) fn to_full_upper(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Upper)
+    }
+
+    // Titlecase mapping is not yet exposed
+    #[allow(dead_code)]
+    pub(crate) fn to_full_title(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Title)
+    }
+
+    // Note: in ICU4C, case folding takes an options bag instead of a locale,
+    // with the only defined option being whether or not to use Turkic (T)
+    // mappings for dotted/dotless i. In ICU4X, we expose a similar locale-free
+    // API for case folding, but internally represent this as a Turkish locale
+    // to simplify shared code.
+    pub(crate) fn to_full_folding(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+    ) -> FullMappingResult {
+        self.full_helper(c, context, locale, MappingKind::Fold)
+    }
+
+    // These constants are used for hardcoded locale-specific foldings.
+    const I_DOT: &'static str = "\u{69}\u{307}";
```

Comment:
suggestion(optional): use 4 hex digits for BMP -- `\u{XXXX}`

Only because these are code points, I'm only used to seeing code points described as U+XXXX with 4 or more hex digits.

FWIW, C++ either allows 4 or 8 hex digits in Unicode char escapes: https://en.cppreference.com/w/cpp/language/escape

Reply:
Will consider for future patch.